pkg/controllers/provisioning/scheduling/topologygroup.go

```diff
@@ -224,17 +243,35 @@ func (t *TopologyGroup) domainMinCount(domains *scheduling.Requirement) int32 {
 	return min
 }
 
+// nolint:gocyclo
 func (t *TopologyGroup) nextDomainAffinity(pod *v1.Pod, podDomains *scheduling.Requirement, nodeDomains *scheduling.Requirement) *scheduling.Requirement {
 	options := scheduling.NewRequirement(podDomains.Key, v1.NodeSelectorOpDoesNotExist)
-	for domain := range t.domains {
-		if podDomains.Has(domain) && t.domains[domain] > 0 {
-			options.Insert(domain)
+
+	// If we are explicitly selecting on specific node domains ("In" requirement),
+	// this is going to be more efficient to iterate through
+	// This is particularly useful when considering the hostname topology key that can have a
+	// lot of t.domains but only a single nodeDomain
+	if nodeDomains.Operator() == v1.NodeSelectorOpIn {
+		for _, domain := range nodeDomains.Values() {
+			if count, ok := t.domains[domain]; podDomains.Has(domain) && ok && count > 0 {
+				options.Insert(domain)
+			}
+		}
+	} else {
+		for domain := range t.domains {
+			if podDomains.Has(domain) && t.domains[domain] > 0 && nodeDomains.Has(domain) {
+				options.Insert(domain)
+			}
 		}
 	}
+	if options.Len() != 0 {
+		return options
+	}
 
-	// If pod is self selecting and no pod has been scheduled yet, we can pick a domain at random to bootstrap scheduling
-
-	if options.Len() == 0 && t.selects(pod) {
+	// If pod is self-selecting and no pod has been scheduled yet OR there is another pod that has scheduled but it's
+	// incompatible with our podDomains, we can pick a domain at random to bootstrap scheduling
+	// This also requires that there not already be some other domain where this a pod in this topology has scheduled
```

Comment:
nit: I think this is a little bit clearer, the last sentence is redundant I think and multiple pods could be scheduled.
```suggestion
	// If pod is self-selecting and no pod has been scheduled yet OR the pods that have scheduled are
	// incompatible with our podDomains, we can pick a domain at random to bootstrap scheduling.
```